File: 卡码/代码随想录/1/05-栈与队列/Python/stackAndQueue/01-有效括号/LeetCode_0020.py
#20. 有效的括号
from typing import *

class Solution:
  def isValid(self, s: str) -> bool:
    if len(s)%2!=0:
      return False
    stack=[]
    for ch in s:
      if ch=="(":
        stack.append(")")
      elif ch=="[":
        stack.append("]")
      elif ch=="{":
        stack.append("}")
      elif len(stack)==0 or stack[-1]!=ch:
        return False
      else:
        stack=stack[:-1]
    return len(stack)==0

File: 卡码/代码随想录/1/05-栈与队列/Java/src/_01_有效括号/LeetCode_0020.java
//20. 有效的括号
package _01_有效括号;

import java.util.Stack;

public class LeetCode_0020 {

}
class Solution {
  public boolean isValid(String s) {
  	if(s.length()%2!=0) return false;
  	Stack<Character> stack = new Stack<>();
  	for(int i=0;i<s.length();i++) {
  		if(s.charAt(i)=='(') stack.push(')');
  		else if(s.charAt(i)=='[') stack.push(']');
  		else if(s.charAt(i)=='{') stack.push('}');
  		else if(stack.empty()||stack.peek()!=s.charAt(i)) return false;
  		else stack.pop();
  	}
  	return stack.empty();
  }
}